build-logic/src/main/kotlin/ktlint-kotlin-common.gradle.kts

```diff
@@ -6,22 +6,38 @@ plugins {
     id("dev.drewhamilton.poko")
 }
 
+val projectLibs = extensions.getByType<VersionCatalogsExtension>().named("libs")
+val javaCompilationVersion = JavaLanguageVersion.of(projectLibs.findVersion("java-compilation").get().requiredVersion)
+val javaTargetVersion = JavaLanguageVersion.of(projectLibs.findVersion("java-target").get().requiredVersion)
+
 kotlin {
     // All modules, the CLI included, must have an explicit API
     explicitApi()
+    jvmToolchain(jdkVersion = javaCompilationVersion.asInt())
 }
 
+tasks.withType<JavaCompile>().configureEach {
+    options.release.set(javaTargetVersion.asInt())
+}
 tasks.withType<KotlinCompile>().configureEach {
-    compilerOptions {
-        jvmTarget.set(JvmTarget.JVM_1_8)
-    }
+    // Convert Java version (e.g. "1.8" or "11") to Kotlin JvmTarget ("8" resp. "11")
+    compilerOptions.jvmTarget.set(JvmTarget.fromTarget(JavaVersion.toVersion(javaTargetVersion).toString()))
 }
 
-// compileJava task and compileKotlin task jvm target compatibility should be set to the same Java version.
-// For some reason, we fallback from toolchain using, see https://github.com/pinterest/ktlint/pull/1787
-tasks.withType<JavaCompile>().configureEach {
-    sourceCompatibility = JavaVersion.VERSION_1_8.toString()
-    targetCompatibility = JavaVersion.VERSION_1_8.toString()
+val requestedJdkVersion = project.findProperty("testJdkVersion")?.toString()?.toInt()
```

Comment:
We should keep using lazy APIs, https://docs.gradle.org/current/javadoc/org/gradle/api/provider/ProviderFactory.html#gradleProperty-java.lang.String-

Reply:
What's the value of doing extra abstraction if the property is consumed immediately, at the configuration time? 

Lazy APIs are useful to take the burden off the single-threaded configuration phase and do more things during task execution time. The documentation: https://docs.gradle.org/current/userguide/lazy_configuration.html#lazy_properties doesn't mention our setup as an example use case to showcase. 
I'd still suggest using simpler version, as long as it's not deprecated, or adding extra complexity comes with _any_ performance benefit 🤷

Reply:
It's the Gradle recommending way, I'd prefer to follow the best practice.

Reply:
> It's the Gradle recommending way

Can you link where Gradle recommends using providers in use cases like this? I agree it's in general recommended to:
1. wire Gradle models
2. wiring Task inputs & outputs
3. Avoiding doing work during configuration phase
None of them are applicable here, and blindly using lazy APIs doesn't bring any value.

I can also link [this](https://docs.gradle.org/current/userguide/lazy_configuration.html#lazy_configuration_faqs): which says:
>Avoid simplifying calls like obj.getProperty().get() and obj.getProperty().set(T) in your code by introducing additional getters and setters.

which is exactly what you'd have to do if you used `gradleProperty()` api 🤷 

An alternative approach to using native Gradle APIs (like I did), is to use [kts delegates](https://docs.gradle.org/current/userguide/kotlin_dsl.html#kotdsl:properties), in this case:`val testJdkVersion: String? by project`, but that's a syntactic sugar and doesn't change behavior (in context of being lazy or not) 🤷 

I'll be happy to be proven wrong, otherwise forcing to use lazy for properties that are consumed only during configuration phase is useless 🤷